_posts/2025-02-28-lattice-bitset.md

```diff
@@ -0,0 +1,801 @@
+post, SSA values) and use those types to optimize. For example, we might see
+the following pseudo-Python snippet:
+
+```python
+a = [...]
+return len(a)
+```
+
+If we know that the type of `a` is `List`, we can optimize the call to
+`len`[^redefining-len] to a direct call to the `list` class's `__len__` method,
+or---even better, since we know about CPython's runtime details---read the
+`ob_size` field directly off the object.
+
+[^redefining-len]: Pretend for now that we can't redefine or shadow `len`
+    because that trickiness is unrelated to this post.
+
+That's great! It will catch a bunch of real-world cases and speed up code.
+Unfortunately, it gives up a little too easily. Consider some silly-ish code
+that either assigns a `str` or a `list` depending on some condition:
+
+```python
+def foo(cond):
+    if cond:
+        a = "..."
+    else:
+        a = [...]
+    return len(a)
+```
+
+Because `a` could have either type, we have to union the type information we
+know into a more general type. Since we don't have  `ListOrString` case in our
+enum (nor should we), we have to pick something more general: `Object`.
+
+That's a bummer, because we as runtime implementors know that both `str` and
+`list` have `__len__` methods that just read the `ob_size` field. Now that we
+have lost so much information in the type union, we have to do a very generic
+call to the `len` function.
+
+To get around this, we can turn the straightforward enum into a *bitset*.
+
+## Bitsets
+
+Bitsets encode set information by assigning each bit in some number (say, a
+64-bit word) a value. If that bit is 1, the value is in the set. Otherwise, it
+isn't.
+
+Here's the enum from earlier, remade as a bitset:
+
+```c
+enum {
+    Int    = 1 << 0,  // 0b001
+    List   = 1 << 1,  // 0b010
+    String = 1 << 2,  // 0b100
+    Object = 7,       // 0b111; catch-all
+};
+```
+
+We have three elements: `Int`, `List`, and `String`. Since we have a bit for
+each object and they can be addressed individually, we can very
+straightforwardly encode the `ListOrString` type using bitwise arithmetic:
+
+```c
+List | String  // 0b110
+```
+
+It doesn't even need to be a basic type in our enum. We can construct it using
+our built-in building blocks.
+
+We have also set our unknown/catch-all/`Object` type as the value with *all*
+bits set, so that if we *or* together all of our known types
+(`Int|List|String`), the unknown-ness falls out naturally.
+
+...but what does the unnamed 0 value represent? If we are thinking in terms of
+sets still (which we should be), then 0 means no bits set, which means it
+represents 0 elements, which means it's the empty set!
+
+We're starting to approach a very useful mathematical structure called a
+*semilattice*.
+
+## Semilattices
+
+I'm not going to quote you any formal definitions because I'm not very mathy,
+but a semilattice is a partial (not total) order of a set with a *join*
+(*union*) operation. It has a top element and a bottom element.
```

Comment:
This is the only paragraph in the post that I had to reread more than twice, probably due to my allergy to formal math. I would have been able to understand if much quicker if it had read like this:

> I’m not going to quote you any formal definitions because I’m not very mathy, but a semilattice is a partial (not total) ordering of sets[^footnote-about-join-op] that places each set below all of its supersets. A semilattice has a top element, which is a superset of all other sets, and a bottom element, which is a subset of all other sets.

I know this may be too simplified to satisfy you, but to me, the abstraction of "join operation" was just confusing, especially since it's never used in this post because we're only ever working with a semilattice of sets.